### ISSUE TITLE
erlang:send/2 to another node can freeze permanently on ARM when TLS is used

### ISSUE THREAD

**Issue by timofey-barmin** [erlang/otp#6309]
**Describe the bug**
In some cases it may happen that erlang:send/2 freezes for good when messages are sent to another node (not locally).
It happens only when distribution over TLS is used AND only when erlang is running on ARM (specifically EC2 c6g).
We did check the vm thread stacktraces and didn't notice anything suspicious: all the threads are basically waiting for an event.

**To Reproduce**
I can reproduce it 100% on erlang 25.0.4 and erlang 24 on a c6g.2xlarge instance with the following script.

[test.erl](https://gist.github.com/timofey-barmin/f36813bd426d3eb9ef7cae6b21f7bfd5)

The script basically spawns several processes that send messages to not existing processes on the other node in the cluster (script assumes there are only two nodes in cluster). 
Each sender sends a message to the remote node and then sends a ping to a local process (the one that started the test). If there is no ping from a sender we assume that erlang:send got stuck.
Hanging also can be confirmed by running net_adm:ping/1 after the test is finished (it will hang as well).

The fact that receiving processes doesn't exist doesn't matter, the bug can be reproduced with existing receivers as well, so I removed receivers for simplicity.

*How to run the test script**

1. Prepare the `ssl_dist_opts` file and make sure it points to correct certificates. I use the following file:

```
[{server,[{certfile, "/home/ec2-user/certs/chain.pem"},
          {keyfile, "/home/ec2-user/certs/pkey.pem"},
          {fail_if_no_peer_cert, false}]},
 {client,[{verify, verify_none}]}].
```
Here is the test certificates that I am using: [certs.zip](https://github.com/erlang/otp/files/9592332/certs.zip) - but I believe the issue doesn't depend on the certs.
 
2. start node 1 with TLS distribution:
`$ erl -proto_dist inet_tls -ssl_dist_optfile "./ssl_dist_opts" -name 'test1@ip-x-x-x-x.us-west-2.compute.internal'`
3. start node 2 with TLS distribution:
`$ erl -proto_dist inet_tls -ssl_dist_optfile "./ssl_dist_opts" -name 'test2@ip-x-x-x-x.us-west-2.compute.internal'`
4. Connect nodes to each other:
```
1> net_adm:ping('test2@ip-x-x-x-x.us-west-2.compute.internal').
pong
```
5. Make sure test.beam is loaded and run the following on either node (this will start 20 senders that will send a message every 10ms, and will stop in 30 second):
```
2> test:start_senders(20, 10, 30000).
...............................................Missing ping from <0.133.0>
.Missing ping from <0.133.0>
.Missing ping from <0.133.0>
.Missing ping from <0.132.0>
.Missing ping from <0.133.0>
...
Missing ping from <0.132.0>
Missing ping from <0.131.0>
Missing ping from <0.130.0>
Missing ping from <0.129.0>
.Missing ping from <0.128.0>

!!!TEST STOP!!!

ok

3> net_adm:ping('test2@ip-x-x-x-x.us-west-2.compute.internal').
<hangs>
```

Each "Missing ping from <pid>" message indicates a hanging send.
After that net_adm:ping (if called) hangs as well.

If you don't see any "Missing ping from" messages, try increasing the number of senders (first param in test:start_senders/3).

**Expected behavior**
erlang:send is not expected to freeze permanently

**Affected versions**
Here is versions that I tried:
Erlang/OTP 25 [erts-13.0.4] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1]
Erlang/OTP 24 [erts-12.1.5] [source-38e24c5703] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1]

**Additional context**
The bug is pretty nasty and blocks work on arm completely. Will appreciate any help. Thanks.

**Comment by mikpe** [comment]
Can you make the description self-contained? Apparently the recipe requires an `ssl_dist_opts` file which you didn't specify the contents of.

**Comment by timofey-barmin** [comment]
Thanks Mikael, I added the ssl_dist_opts file. I also added certs.zip with certificates that I was using. Just in case.

**Comment by IngelaAndin** [comment]
If this is some kind of deadlock in Erlang it would be helpful if you could generate an erl-crashdump when it happens. `erlang:halt("Freeze").`

**Comment by mikpe** [comment]
I don't have easy access to a Graviton cluster (we're still on x86-64 in prod) but I tried the recipe with two nodes on the same server, OTP-25.0.4, on x86-64, sparc64, and armv7l, all running Linux, but none of them would reproduce the issue. Given that you see it also with OTP-24 I don't think the arm64 jit can be involved, it's either in generic code or something outside of the Erlang VM proper.

A `gdb` dump of the hanging thread (at the final `net_adm:ping/1`) might be useful.

**Comment by timofey-barmin** [comment]
It seems like the issue is happening because tls_sender is not receiving the dist_data message here:

https://github.com/erlang/otp/blob/master/lib/ssl/src/tls_sender.erl#L322

Basically it looks like sometimes this message is not sent for unknown reason. Tls_sender never tries to send data to the socket. Buffer gets filled, and eventually senders get blocked.
If I send the `dist_data` message to the tls_sender process manually - everything starts working magically (until the issue reproduces again). So it looks like a some kind of synchronization issue, or race in the erlang vm.  

For example, something like the following can be used as a work-around hypothetically:
`spawn(fun R() -> [P ! dist_data || {_, P} <- erlang:system_info(dist_ctrl)], timer:sleep(1000), R() end).`

It periodically sends messages to all dist tls senders, which unblocks them if the issue happens.

I also experimented a bit with it and here are some facts that can be useful:
1. If I set the number of schedulers to 1, the issue never happens.
2. Sometimes it just works with no problems at all. I need to restart the node in this case in order to reproduce the issue again.
3. It doesn't depend on jit. I tested it with and without jit - same behavior. 

Here is an [erl_crash.zip](https://github.com/erlang/otp/files/9603136/erl_crash.zip) that was taken while the issue was happening. Specifically the process <0.134.0> was hanging in erlang:send/2.

**Comment by timofey-barmin** [comment]
I understand that this is probably a huge overkill, and most likely this patch will make distribution much slower, but as an experiment I tried it. It seems like it completely fixes the issue for me:

```
diff --git a/erts/emulator/beam/dist.c b/erts/emulator/beam/dist.c
index 464fd528e7..cb4082630b 100644
--- a/erts/emulator/beam/dist.c
+++ b/erts/emulator/beam/dist.c
@@ -3991,6 +3991,7 @@ dist_ctrl_get_data_notification_1(BIF_ALIST_1)

     ASSERT(dep->cid == BIF_P->common.id);

+    erts_mtx_lock(&dep->qlock);
     qflgs = erts_atomic32_read_acqb(&dep->qflgs);

     if (!(qflgs & ERTS_DE_QFLG_REQ_INFO)) {
@@ -4015,6 +4016,7 @@ dist_ctrl_get_data_notification_1(BIF_ALIST_1)
     }
     /* else: Already requested... */

+    erts_mtx_unlock(&dep->qlock);
     erts_de_runlock(dep);

     if (is_internal_pid(receiver))
```
I still don't understand this code well though. So I am rather guessing at this point. Please let me know if it doesn't make any sense.
Any help from those who knows this code is appreciated. Thanks!

**Comment by timofey-barmin** [comment]
@IngelaAndin @mikpe 
Will it help you if we get you an AWS ARM box to try and repro the issue on? This way you will have a reproducible case to investigate.
If it makes sense, please let me know your id_rsa.pub (you can put it here or you can email it to me). I can give you an ssh access to the instance then.
My email is timofey.barmin@couchbase.com.
Alternatively we can email you credentials to access the instance.
Thanks

**Comment by rickard-green** [comment]
@timofey-barmin I think I've found the issue. I'll soon make a PR which you can test.

**Comment by rickard-green** [comment]
@timofey-barmin Please try #6324. The branch in the pull request is based on OTP 22.3.4 (in order to be able to reuse it for patches on multiple releases). It will cleanly merge to any OTP version greater than 22.3.4, so you should be able to test it on any version you want.

I've tried to recreate this issue on my M1 Mac (arm64), but without succeeding.

**Comment by timofey-barmin** [comment]
@rickard-green 
I applied this patch to the maint-25 branch and performed the test multiple times. It seems to fix the issue completely for me. Very impressive work!

Quick question: Do I understand correctly that this bug may affect other platforms as well? basically any platform that supports different kinds of barriers should be affected (I have to admit though: I don't know what those platforms are).

Also while I am still trying to understand the patch I noticed that the following one line change alone also makes the issue not reproducible for me:
```
diff --git a/erts/emulator/beam/dist.c b/erts/emulator/beam/dist.c
index 464fd528e7..bae6aec67a 100644
--- a/erts/emulator/beam/dist.c
+++ b/erts/emulator/beam/dist.c
@@ -3383,7 +3383,7 @@ erts_dsig_send(ErtsDSigSendContext *ctx)
                        (!ctx->no_trap && !ctx->no_suspend));

                erts_mtx_lock(&dep->qlock);
-               qsize = erts_atomic_add_read_nob(&dep->qsize, (erts_aint_t) obsz);
+               qsize = erts_atomic_add_read_mb(&dep->qsize, (erts_aint_t) obsz);
                 ASSERT(qsize >= obsz);
                 qflgs = erts_atomic32_read_nob(&dep->qflgs);
                if (!(qflgs & ERTS_DE_QFLG_BUSY) && qsize >= erts_dist_buf_busy_limit) {

```

**Comment by rickard-green** [comment]
> @rickard-green I applied this patch to the maint-25 branch and performed the test multiple times. It seems to fix the issue completely for me. Very impressive work!
> 

Thanks to [you zooming in on the location of the problematic code](https://github.com/erlang/otp/issues/6309#issuecomment-1253001157) it was fairly easy to find the problem.

> Quick question: Do I understand correctly that this bug may affect other platforms as well? basically any platform that supports different kinds of barriers should be affected (I have to admit though: I don't know what those platforms are).
> 

Yes, the only platforms I know, on the top of my head, that aren't effected by this are x86 and x86_64. This since `erts_atomic_add_read_nob()` cannot be implemented without a memory barrier, and `erts_atomic32_read_nob()`s cannot be reordered with each other on x86/x86_64.

There might however be other hardware that theoretically is affected since the memory model that it use allows the problematic reorderings, but in practice the actual hardware is stricter than the memory model, so it is not effected or at least not yet effected. That might be the reason for me not being able to reproduce it on the M1 Mac.

> Also while I am still trying to understand the patch I noticed that the following one line change alone also makes the issue not reproducible for me:
> 
> ```
> diff --git a/erts/emulator/beam/dist.c b/erts/emulator/beam/dist.c
> index 464fd528e7..bae6aec67a 100644
> --- a/erts/emulator/beam/dist.c
> +++ b/erts/emulator/beam/dist.c
> @@ -3383,7 +3383,7 @@ erts_dsig_send(ErtsDSigSendContext *ctx)
>                         (!ctx->no_trap && !ctx->no_suspend));
> 
>                 erts_mtx_lock(&dep->qlock);
> -               qsize = erts_atomic_add_read_nob(&dep->qsize, (erts_aint_t) obsz);
> +               qsize = erts_atomic_add_read_mb(&dep->qsize, (erts_aint_t) obsz);
>                  ASSERT(qsize >= obsz);
>                  qflgs = erts_atomic32_read_nob(&dep->qflgs);
>                 if (!(qflgs & ERTS_DE_QFLG_BUSY) && qsize >= erts_dist_buf_busy_limit) {
> ```

Without the memory barrier on the `add-read` of `qsize`, the following `read` of `qflgs` could be re-ordered with it, so we didn't see that `ERTS_DE_QFLG_REQ_INFO` had been set. A similar exists [here](https://github.com/erlang/otp/pull/6324/files#diff-d256ca218de9be9017823817bd9953a7f3000a190a44dd34956e446b8d68b478R4146) where the `read`s of `qsize` and `qflgs` could be reordered. This one is however much harder to hit, since it may only occur once when the connection goes up.

**Comment by daverigby** [comment]
Out of curiosity, does ThreadSanitizer (https://github.com/google/sanitizers/wiki/ThreadSanitizerCppManual) catch this issue? In _theory_ if erlang is built to use GCC builtin atomics then TSan should be able to flag this kind of issue...

**Comment by rickard-green** [comment]
@daverigby Unfortunately we cannot use GCC `__atomic*` API out of the box since our atomic API isn't compatible with it (it was made years before the `__atomic*` one). In order to be able to do that we would have to go through all uses of atomics in the emulator and change uses at various places which will be a very large job. Doing this might, however, in the long run be worth it, if we could catch bugs like this with tools like thread sanitizer. This is however not anything that we have scheduled yet.

**Comment by daverigby** [comment]
@rickard-green Thanks for the response. 

I had assumed that if the erlang atomic API was ultimately implemented in terms of GCC atomics (which it appeared to be when I looked - but I could be mistaken) then TSan would still be of value. 

However TSan's checks are based on the C11 memory model, so if erlang has defined it's own model then I can imagine you could see a lot of false positive issues...

**Comment by rickard-green** [comment]
The fix in #6324 was released in [OTP 25.1.1](https://www.erlang.org/patches/otp-25.1.1) earlier today. The fix will also be included in patches on OTP 23 and OTP 24 which probably will be released later this week.
